allow not starting tsserver immediately when starting tide

lddubeau opened the pull request:
This is the companion WIP to issue #294 

I've been using this for a few weeks now without running into actual issues (commit dates are deceptive: I used it before the code was committed) so I figure I should put it up for public view.

Going into a buffer that does not have `tsserver` started causes some things to not work but *that's expected*. When I finally start a `tsserver` manually it works just as well as if the server had been started automatically.

ananthakumaran commented on tide.el:
```diff
@@ -1869,28 +1898,15 @@ code-analysis."
   (set (make-local-variable 'imenu-create-index-function)
        'tide-imenu-index)
 
-  (tide-configure-buffer)
```
any reason for removing the configure-buffer call for immediate mode? Note that we can't depend on `hack-local-variables-hook` see https://github.com/ananthakumaran/tide/commit/fbdef5d7af25786d2eddd691a8d6a00225c4202d

lddubeau replied:
Ah, yes. When someone creates a new TS buffer, then `tide-configure-buffer` has never been called to configure this buffer. If `hack-local-variables-hook` fails to run, then the buffer won't be configured at all. I did not run into problems because presumably, with my configuration, `hack-local-variables-hook` always runs and took care of configuring the buffer.  I'll fix that.

ananthakumaran commented:
rest of the changes looks straight forward to me.

lddubeau commented:
I've added a [third commit](https://github.com/ananthakumaran/tide/pull/303/commits/26c208a24cbfcd7516c4fade1aec5c7578090080) that takes care of the issue pointed out in the earlier [code review](https://github.com/ananthakumaran/tide/pull/303#pullrequestreview-202084433).

And I've rebased onto the current `master`.

ananthakumaran commented on tide.el:
```diff
@@ -1871,28 +1895,27 @@ code-analysis."
+      ;;
+      (tide-configure-buffer)
+    (when (eq tide-tsserver-start-method 'immediate)
+      (tide-start-server)))
```
when we start the server for the first time, do we expect the tide-start-server to call configure-buffer? I am assuming that would miss the current buffer because tide-mode is not yet activated?

lddubeau replied:
Yeah, there's a problem there. I think at this point since we're discussing details rather than general objections to the approach, I should just get the tests together (which would most likely have revealed the problem here).

lddubeau commented:
I've added formal tests and fixed the execution path that resulted in a buffer not getting configured.

josteink commented on tide-tests.el:
```diff
@@ -219,6 +266,64 @@ a test failure."
+        (should (eq (tide-current-server) nil))
+        (should (not (spy-called configure-spy)))
+        (tide-restart-server)
+        (should (spy-called configure-spy))
```
Should we have a `(should (not (eq (tide-current-server) nil)))` here for consistency?

josteink commented:
I've looked over the overall changes and I see nothing *obviously* wrong, nor any changes which looks particularly dangerous.

If you say this has been working for you "for months", I see no reason for not merging this.

lddubeau commented:
@josteink Thanks. I'm going to merge this now.
